core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MergeProjectWithValues.java

```diff
@@ -0,0 +1,201 @@
+
+    public MergeProjectWithValues(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isMergeProjectWithValues(session);
+    }
+
+    @Override
+    public Result apply(ProjectNode node, Captures captures, Context context)
+    {
+        ValuesNode valuesNode = captures.get(VALUES);
+
+        // handle projection which prunes all symbols
+        if (node.getOutputSymbols().isEmpty()) {
+            return Result.ofPlanNode(new ValuesNode(valuesNode.getId(), valuesNode.getRowCount()));
+        }
+
+        // fix iteration order over ProjectNode's assignments
+        Map<Symbol, Expression> assignments = ImmutableMap.copyOf(node.getAssignments().getMap());
+
+        // handle values with no output symbols
+        if (valuesNode.getOutputSymbols().isEmpty()) {
+            ImmutableList.Builder<Expression> projectedRows = ImmutableList.builder();
+            for (int i = 0; i < valuesNode.getRowCount(); i++) {
+                projectedRows.add(new Row(ImmutableList.copyOf(assignments.values())));
+            }
+            return Result.ofPlanNode(new ValuesNode(valuesNode.getId(), node.getOutputSymbols(), projectedRows.build()));
+        }
+
+        // do not proceed if ValuesNode contains a non-deterministic expression and it is referenced more than once by the projection
+        Set<Symbol> nonDeterministicValuesOutputs = new HashSet<>();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            Row row = (Row) rowExpression;
+            for (int i = 0; i < valuesNode.getOutputSymbols().size(); i++) {
+                if (!isDeterministic(row.getItems().get(i), metadata)) {
+                    nonDeterministicValuesOutputs.add(valuesNode.getOutputSymbols().get(i));
+                }
+            }
+        }
+        Set<Symbol> multipleReferencedSymbols = node.getAssignments().getExpressions().stream()
+                .map(SymbolsExtractor::extractAll)
+                .flatMap(Collection::stream)
+                .collect(toImmutableMultiset()).entrySet().stream()
+                .filter(entry -> entry.getCount() > 1)
+                .map(Multiset.Entry::getElement)
+                .collect(toImmutableSet());
+        if (!Sets.intersection(nonDeterministicValuesOutputs, multipleReferencedSymbols).isEmpty()) {
+            return Result.empty();
+        }
+
+        // inline values expressions into projection's assignments
+        RowType projectedRowType = RowType.anonymous(assignments.keySet().stream()
+                .map(symbol -> context.getSymbolAllocator().getTypes().get(symbol))
+                .collect(toImmutableList()));
```

Comment:
This multiline argument is a little hard to read. At first glance, it's confusing as to whether the `.map` applies to the output of `RowType.anonymous`.

Format it like this:

```java
RowType projectedRowType = RowType.anonymous(
    assignments.keySet().stream()
        .map(symbol -> context.getSymbolAllocator().getTypes().get(symbol))
        .collect(toImmutableList()));
```

Reply:
Also, the lambda can be replaced with a method reference:

```java
.map(context.getSymbolAllocator().getTypes()::get)
```